rubin_scheduler/utils/consdb.py

```diff
@@ -0,0 +1,1057 @@
+# will have differenc schema in the consdb, so make a base class with
+# expected common functionality, and use derived classes for
+# instrument-specific elements.
+
+# Separate different pieces of metadata (columns in a table of visits) into
+# different methods to make it easy to individually override them by
+# instrument.
+
+
+# cached_properties are a way to compute values when and only when needed
+# but these won't update if the date is changed. In my expected use, that's
+# fine; the user should just instantiate another ConsDBVisits instance if
+# they want a different database or date. But, what if someone does?
+# To avoid this, make it a frozen dataclass, which prevents members from
+# being updated.
+@dataclass(frozen=True)
+class ConsDBVisits(ABC):
+    """Visits returned from a query to the consdb.
+
+    Parameters
+    ----------
+    day_obs : `str`
+        The date for which to query the database.
+    url : `str`
+        The connection string from the database.
+    """
+
+    day_obs: str | int
+    url: str = "postgresql://usdf@usdf-summitdb.slac.stanford.edu:5432/exposurelog"
+
+    @property
+    @abstractmethod
+    def instrument(self) -> str:
+        """The instrument.
+
+        Returns
+        -------
+        instrument : `str`
+            The instrument.
+        """
+        # Subclasses should return a string constant specifying the relevant
+        # instrument.
+        # Specified as an property rather than an attribute so it can be
+        # abstract.
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def num_exposures(self) -> int:
+        """The number of snaps per visit.
+
+        Returns
+        -------
+        num_exposures : `int`
+            The number of snaps per visit.
+        """
+        # Subclasses should return a string constant specifying the relevant
+        # instrument.
+        # Specified as an property rather than an attribute so it can be
+        # abstract.
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def telescope(self) -> str:
+        """The telescope.
+
+        Returns
+        -------
+        telescope : `str`
+            The telescope.
+        """
+        # Subclasses should return a constant specifying the relevant
+        # instrument.
+        # Specified as an property rather than an attribute so it can be
+        # abstract.
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def gain(self) -> pd.Series:
+        """The gain.
+
+        Returns
+        -------
+        gain : `pd.Series`
+            The gain.
+        """
+        # Subclasses should return a constant specifying the relevant
+        # instrument.
+        # Specified as an property rather than an attribute so it can be
+        # abstract.
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def pixel_scale(self) -> pd.Series:
```

Comment:
Yes, at present they are hard-coded because they are not in consdb. I separated them out into their own properties rather than just using constants to make it easy to update the class if they ever become available, or even to support a mix of instruments for which they are available for some and not others.